File: 98-validate-binary-search-tree/98-validate-binary-search-tree.java

class Solution {
    public boolean isValidBST(TreeNode root) {
  return isValidBST(root, Double.NEGATIVE_INFINITY, Double.POSITIVE_INFINITY);    
}
 
public boolean isValidBST(TreeNode a, double min, double max){
    if(a==null) 
        return true;
 
    if(a.val <= min || a.val >= max)
        return false;
 
    return isValidBST(a.left, min, a.val) && isValidBST(a.right, a.val, max);
    }
}

File: GO-Leet/98. Validate Binary Search Tree.go
package main

func isValidBST(root *TreeNode) bool {
	return tree(root, nil, nil)
}

func tree(root, min, max *TreeNode) bool {
	if root == nil {
		return true
	}

	if min != nil && root.Val <= min.Val {
		return false
	}
	if max != nil && root.Val >= max.Val {
		return false
	}

	return tree(root.Left, min, root) && tree(root.Right, root, max)
}
